### DISCUSSION TITLE
Difficulty providing container types as bound parameters

### DISCUSSION THREAD

**Discussion by skreborn** [cornucopia-rs/cornucopia#139] [General]
At the moment, it's difficult to provide container types such as wrappers (`Option`) and collections (`Vec`) as bound parameters, because everything _has_ to be converted to a borrowed version. This is most annoying in the case of a `Vec<String>` that has to be converted to a `&[&str]` before binding. It only gets worse when I wrap it in an `Option`.

This introduces some awful code duplication because I constantly have to fight the borrow checker, when `postgres` would happily accept my `Option<Vec<String>>` as-is (as far as  I can tell at least).

I'm unsure what would be the proper solution to this problem. At least for `String` and `&str` values, `Cow` could be used, as `postgres` has `ToSql` implemented for it.

Another approach could be to create a separate borrowed and owned parameter struct, much like when returning results. This way, the library user could make the conscious decision to lean either way - whichever works for them.

**Comment by LouisGariepy** [comment]
Thanks for reporting. There indeed seems to be a lack of ergonomics here. I'll think this through and open a PR soon!

**Reply by skreborn** [reply]
Sounds good. Let me know if there's anything I can do to help.
